### PR TITLE
Power FSM boilerplate & toggle between HP/LP 

### PR DESCRIPTION
**Pull request by fustilio**
Requires the use of [FunctionFsm arduino library](https://github.com/JRVeale/function-fsm/)
image
on_enter, on_exit functions occur once whereas on_state functions execute on every fsm update.

Introduces two power states: `LOW_POWER `and `HIGH_POWER`
Transit from `LOW_POWER` to `HIGH_POWER` when device is charging **or** when device is no longer charging but battery level is at 100 (this indicates that the device is plugged in but at full capacity so not charging)
Transit from `HIGH_POWER` to `LOW_POWER` when device is no longer charging **and** when device battery level is less than 100 (this and statement prevents transition from `HIGH_POWER` to `LOW_POWER` states when device has completed charging)

### CODE REVIEW

File: alpha/io.h [hunk 1]
```diff
@@ -1,7 +1,6 @@
 #include <stdint.h>
 #include <FunctionalInterrupt.h>
 #include <Arduino.h>
-#include "hal.h"
```
Comment (fustilio) [inline alpha/io.h:4]: remove unnecessary include statement

File: alpha/fsm.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
+#include "fsm.h"
+#include "hal.h"
+
+_TS_FSM TS_FSM;
```
Comment (jamesi8086) [inline alpha/fsm.cpp]: just realized this is a power controller fsm, so it should be like TS_Power

File: alpha/fsm.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+{ 
+  Serial.println("- Enter LP -"); 
+  current_state = TS_PowerState::LOW_POWER;
+}
+
+void _TS_FSM::lp_on()
+{
+  if (millis() - last_print > interval)
+  {
+    Serial.println("STATE: LOW POWER");
```
Comment (jamesi8086) [inline alpha/fsm.cpp]: can we convert all Serial.println to log_i
log_e for errors
log_w for warnings

File: alpha/fsm.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  fsm.add_transition(&state_lp, &state_hp, TOGGLE_SWITCH, [this]() { lp_on_trans_hp(); } );
+  fsm.add_transition(&state_hp, &state_lp, TOGGLE_SWITCH, [this]() { hp_on_trans_lp(); } );
+}
+
+void _TS_FSM::update()
+{
+  fsm.run_machine();
+}
+
+void _TS_FSM::toggleFsm()
```
Comment (jamesi8086) [inline alpha/fsm.cpp]: toggle_fsm

File: alpha/ui.cpp [hunk 5]
```diff
@@ -213,6 +214,17 @@ void _TS_UI::task(void* parameter)
       TS_HAL.lcd_printf("Power long");
     }
 
+    // TODO: To be removed, for testing purposes only
+    if (TS_FSM.current_state == TS_PowerState::LOW_POWER) 
+    {
+      TS_HAL.lcd_cursor(0, 0);
+      TS_HAL.lcd_printf("State: Low Power \n");
+    } else if (TS_FSM.current_state == TS_PowerState::HIGH_POWER)
```
Comment (jamesi8086) [inline alpha/ui.cpp]: nit: fix style

File: alpha/ui.cpp [hunk 6]
```diff
@@ -213,6 +214,17 @@ void _TS_UI::task(void* parameter)
       TS_HAL.lcd_printf("Power long");
     }
 
+    // TODO: To be removed, for testing purposes only
+    if (TS_FSM.current_state == TS_PowerState::LOW_POWER)
```
Comment (jamesi8086) [inline alpha/ui.cpp]: TODO: icon to show power state on ui

**Review by jamesi8086 — APPROVED** [review]
except some nitpicks, looks good

File: alpha/fsm.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+enum TS_PowerState
+{
+  LOW_POWER,
+  HIGH_POWER,
+};
+
+class _TS_FSM
+{
+  public:
```
Comment (jamesi8086) [inline alpha/fsm.h]: hmm, is there an accessor for another task to check what power state this is in ?
main radio task needs to check to switch between modes e.g. wifi or not

**Comment by fustilio** [comment]
Made changes according to comments, state accessor at TS_POWER.get_state()

**Comment by jamesi8086** [comment]
lgtm except that, rename the whole file away from fsm.cpp/h to power.cpp/h

**Pull request merged**